### DISCUSSION TITLE
implement URDNA2015

### DISCUSSION THREAD

**Discussion by rwst** [RDFLib/rdflib#1545] [Ideas]
A quick skim of code/issues didn't turn up anything. Please implement the [RDF Dataset Normalization Algorithm](https://json-ld.github.io/normalization/). See also https://github.com/digitalbazaar/rdf-canonize.

**Comment by nicholascar** [comment]
Looks useful @rwst, are you proposing to implement this yourself? If you need this algorithm for any purpose, please do so and we will review.

**Comment by rwst** [comment]
I'm tempted, but not right now.

**Comment by ghost** [comment]
> Please implement the [RDF Dataset Normalization Algorithm](https://json-ld.github.io/normalization/). See also https://github.com/digitalbazaar/rdf-canonize.

URL has changed to  [RDF Dataset Canonicalization](https://json-ld.github.io/rdf-dataset-canonicalization/spec/), js reference implementation remains the same: https://github.com/digitalbazaar/rdf-canonize

**Comment by ff137** [comment]
Moved again: https://www.w3.org/TR/rdf-canon/

Does anyone know the current state of normalization / canonicalization support in RDFLib?

**Reply by mielvds** [reply]
There are old remnants of this in https://github.com/digitalbazaar/pyld/blob/58ea0ea9e64c87b5ab8da78016b1cae170700ddf/lib/pyld/jsonld.py#L248, back when it was still part of the JSON-LD spec. This (and its tests) should be taken out of PyLD and have place of its own in the RDF ecosystem. Of course, it should also be made compliant with the recommendation.

**Reply by mielvds** [reply]
see https://github.com/digitalbazaar/pyld/issues/220 btw

**Comment by mielvds** [comment]
RDFC1.0 is implemented using rdflib in PyLD, see https://github.com/digitalbazaar/pyld/pull/245. This will probably move to rdflib, stay tuned

**Reply by gjhiggins** [reply]
> RDFC1.0 is implemented using rdflib in PyLD

<Doh!> I totally missed this. For my own entertainment, I idly translated the Digital Bazaar Javascript into Python and rolled it up into an RDFLib Serializer plugin which I coincidentally parked [on codeberg](https://codeberg.org/gjhiggins/rdflib-rdna) about a week ago. I've been sat on it a while 'cos it's failing a dozen or so of the W3C test cases and I haven't got round to looking into it.

Restructuring it as a plugin necessitates a reduction of options (such as setting the prefix) but, convenience has a strong pull these days ...
```python
from rdflib import Dataset

data = (
    '_:b4 <http://purl.org/dc/terms/publisher> "Harry" .'  # triple
    '_:b3 <http://xmlns.com/foaf/0.1/name> "Harry" _:b4 .'  # quad
)

expected = (
    '_:c14n0 <http://xmlns.com/foaf/0.1/name> "Harry" _:c14n1 .'
    '_:c14n1 <http://purl.org/dc/terms/publisher> "Harry" .'
)

def test_rdna():
    # Instantiate a Dataset
    ds = Dataset()

    # Parse the N-Quads as usual
    ds.parse(data=data, format="nquads")

    # Serialize as canonical
    canonicalized = ds.serialize(format="rdna")

    assert canonicalized == expected

    # Closedown
    ds.close()
```

**Reply by mielvds** [reply]
Cool! It's nice to see how easy this has become. I hope we can soon spend our collective efforts (or tokens) on improving what's there instead of creating out own ;) FWIW, the branch in the PR passes all w3c tests. Your test would be:

```
from pyld.canon import RDFC10

data = (
    '_:b4 <http://purl.org/dc/terms/publisher> "Harry" .'  # triple
    '_:b3 <http://xmlns.com/foaf/0.1/name> "Harry" _:b4 .'  # quad
)

expected = (
    '_:c14n0 <http://xmlns.com/foaf/0.1/name> "Harry" _:c14n1 .'
    '_:c14n1 <http://purl.org/dc/terms/publisher> "Harry" .'
)

def test_rdna():
    # Serialize as canonical
    canonicalized = RDFC10().main(data)

    assert canonicalized == expected
```

**Reply by gjhiggins** [reply]
> Cool! It's nice to see how easy this has become. I hope we can soon spend our collective efforts (or tokens) on improving what's there instead of creating out own ;) 

Yes, structuring an addition as a plugin is fairly straightforward and it has the advantage of not adding to the maintenance burden of core RDFlib. Out of personal interest, I made a cursory comparison of our implementations and have  been able to glean some ideas for improvements in mine, based on yours (I'm a cognitive psychologist by discipline and have "picked up" programming along the way, so I'm always mildly interested in improving my style and to be able to make a direct comparision with a very well-written implementation is both rare and extremely useful personally, so thank you for that opportunity).

As regards to ploughing my own path, my excuse is that I made the translation 4 years ago in response to the OP here and regarded the test failures as evidence of "close but no cigar", so parked it for the interim. I did note that your branch passed all the w3C tests and am hopeful that I'll be able to nail down where I went wrong in my translation.

Also, I wrote (above) incorrectly ...
> Restructuring it as a plugin necessitates a reduction of options

because I keep forgetting that RDFLib plugin classes typically accept an [`**args`](https://github.com/RDFLib/rdflib/blob/0fdac395078b81313f30eeab74b1a18eb7bb4db6/rdflib/serializer.py#L37) parameter binding, I need to develop my approach further in order to enable the specification of options via that mechanism.

**Reply by gjhiggins** [reply]
>am hopeful that I'll be able to nail down where I went wrong in my translation.

I decided to just junk it and re-do it based on a direct adaptation of mielvds’s pyld implementation instead - see [rdflib-rdna](https://codeberg.org/gjhiggins/rflib-rdna), as he kindly hinted.

> because I keep forgetting that RDFLib plugin classes typically accept an [`**args`](https://github.com/RDFLib/rdflib/blob/0fdac395078b81313f30eeab74b1a18eb7bb4db6/rdflib/serializer.py#L37) parameter binding, I need to develop my approach further in order to enable the specification of options via that mechanism.

Managed to do that. `RDNAGraphCanonicalizer.canonicalize()` accepts `**kwargs`. By default it returns a generator of canonicalized quads, given `as_dataset=True` it returns a canonicalzed Dataset. Other `kwarg` options that are recognized are:

- `hash_algorithm="ALGO"` - anything that appears in `hashlib.algorithms_guaranteed`.
- `normalize_literals=False|True`
- `prefix=str`

The serializer uses the same kwargs where relevant (normalize_literals, prefix, hash_algorithm) but I've yet to write tests.

Given data of:
```python
data = (
    '_:b4 <http://purl.org/dc/terms/publisher> "Harry" .\n'
    '_:b3 <http://xmlns.com/foaf/0.1/name> "Harry" _:b4 .\n'
)
```
usage is:
```python
ds = Dataset()
ds.parse(data=data, format="nquads")

# Returns a generator of canonicalized quads
canonicalized_nquads = RDNAGraphCanonicalizer(ds).canonicalize()

# canonicalised nquads are reliably sorted
assert next(canonicalized_nquads) == (
    BNode('c14n0'),
    DCTERMS.publisher,
    Literal("Harry"),
    ds.default_graph.identifier
)

assert next(canonicalized_nquads) == (
    BNode('c14n1'),
    FOAF.name,
    Literal("Harry"),
    BNode('c14n0')
)

# Specify `as_dataset=True` as a parameter binding, gets detected
# via inspection of **kwargs
ds2 = RDNAGraphCanonicalizer(ds).canonicalize(as_dataset=True)

# Default graphs are isomorphic
assert ds.default_graph.isomorphic(ds2.default_graph)

# As are the respective named graphs
assert ds.graph(BNode('b4')).isomorphic(ds2.graph('c14n0'))
```

which is nice.

Cheers,

Graham
